Automattic/jetpack
# Lazy Images: Use W3C IntersectionObserver polyfill

Fixes #8821

When we first worked on Lazy Images, we were using the W3TC IntersectionObserver polyfill, but near the end, we moved to another polyfill because the W3TC version had conflicting licenses. Specifically, in one case they were using an Apache 2.0 license which isn't GPL compatible. You can see more discussion about that in #8093.

As of https://github.com/w3c/IntersectionObserver/commit/4a9a908f88f6d38be068b9dd523f86d281c5398b and https://github.com/w3c/IntersectionObserver/commit/b025731ab0f4bdbd444ad1b0917b8d5991774548, they have updated their licenses to "W3C SOFTWARE AND DOCUMENT NOTICE AND LICENSE", which is GPL compatible!

Since this is the one being worked on by W3C, I imagine that it's the most up-to-date and well tested. Using this polyfill fixes the Twitter embed issue reported in #8821.

We should do regression testing in other browsers before shipping.

Note: This is a large PR, but it is essentially copying and pasting a new polyfill in. 😄 

To test:

- Checkout branch
- run `yarn build` to build the minified files
- Ensure that license appears in the minified lazy images JavaScript files
- Create several posts with images
- Turn lazy images on
- Ensure that images load lazily as expected
- Set `define( 'SCRIPT_DEBUG', true );` in `wp-config.php` or an integration plugin, which will use the non-minified scripts
- Ensure that images load lazily as expected
